src/Host/Broker/Impl/Sessions/Session.cs

```diff
@@ -104,108 +109,57 @@ public class Session {
-                    RedirectStandardInput = true,
-                    RedirectStandardOutput = true,
-                    RedirectStandardError = true,
-                    LoadUserProfile = true
-                };
-
-
-                psi.EnvironmentVariables["R_HOME"] = shortHome.ToString();
-                psi.EnvironmentVariables["PATH"] = Interpreter.Info.BinPath + ";" + Environment.GetEnvironmentVariable("PATH");
-
-                psi.WorkingDirectory = Path.GetDirectoryName(rhostExePath);
-
-                _process = new Process {
-                    StartInfo = psi,
-                    EnableRaisingEvents = true,
-                };
-
-                _process.ErrorDataReceived += (sender, e) => {
-                    var process = (Process)sender;
-                    outputLogger?.LogTrace(Resources.Trace_ErrorDataReceived, process.Id, e.Data);
-                };
-
-                _process.Exited += delegate {
-                    _hostEnd?.Dispose();
-                    _hostEnd = null;
-                    State = SessionState.Terminated;
-                };
-
-                _sessionLogger.LogInformation(Resources.Info_StartingRHost, Id, User.Name, rhostExePath, arguments);
-                try {
-                    StartSession();
-                } catch (Exception ex) {
-                    _sessionLogger.LogError(Resources.Error_RHostFailedToStart, ex.Message);
-                    throw;
-                }
+                eb = Win32EnvironmentBlock.Create(WindowsIdentity.GetCurrent().Token);
+            }
 
-                _sessionLogger.LogInformation(Resources.Info_StartedRHost, Id, User.Name);
+            // add additional variables to the environment block
+            eb["R_HOME"] = shortHome.ToString();
+            _sessionLogger.LogTrace(Resources.Trace_EnvironmentVariable, "R_HOME", eb["R_HOME"]);
+            eb["PATH"] = Invariant($"{Interpreter.Info.BinPath};{eb["PATH"]}");
+            _sessionLogger.LogTrace(Resources.Trace_EnvironmentVariable, "PATH", eb["PATH"]);
 
-                _process.BeginErrorReadLine();
 
-                stdin = _process.StandardInput.BaseStream;
-                stdout = _process.StandardOutput.BaseStream;
+            _sessionLogger.LogInformation(Resources.Info_StartingRHost, Id, User.Name, rhostExePath, arguments);
+            using (Win32NativeEnvironmentBlock nativeEnv = eb.GetNativeEnvironmentBlock()) {
+                int pid = 0;
+                if (loggedOnUser) {
+                    pid = Win32Process.StartProcessAsUser(useridentity, rhostExePath, arguments, Path.GetDirectoryName(rhostExePath), nativeEnv, out stdin, out stdout, out stderror);
+                }else {
+                    pid = Win32Process.StartProcessAsUser(null, rhostExePath, arguments, Path.GetDirectoryName(rhostExePath), nativeEnv, out stdin, out stdout, out stderror);
+                }
+                _process = Process.GetProcessById(pid);
```

Comment:
There is a race condition here, since the process might crash before we try to look it up. 

We can handle `ArgumentException`, but there's still a potential race if a new process starts concurrently, and reuses the same PID. So it seems that the only safe way to handle this is to not use `Process` at all.

It looks like at this point, the only things that we're actually using it for is checking if the process is still running, and registering a handler for when it exists. BTW, the latter might also be a race now, since we register the handler after starting the process, not before; I'm not sure what happens if process terminates (e.g. crashes) before we register the handler, but I wouldn't be surprised if the handler doesn't get called.

So, I think we should just bite the bullet and move completely to Win32 here for dealing with processes. Stash away the process handle, not just PID; and use `WaitOne` and `ThreadPool.RegisterWaitForSingleObject` to detect process termination via the handle.